### PR TITLE
[#181] Common base class to eliminate the code duplication for exchan…

### PR DESCRIPTION
**Pull request by sysexcontrol**
…ge objects of device registration and credentials API.
Last refactoring PR of this issue.
(Two minor javadoc fixes contained as well)

Signed-off-by: Karsten Frank <Karsten.Frank@bosch-si.com>

### CODE REVIEW

**Comment by sysexcontrol** [comment]
Javadoc fixes removed again, since they were corrected in parallel on master :+1:

File: service-base/src/main/java/org/eclipse/hono/service/registration/BaseRegistrationService.java [hunk 1]
```diff
@@ -139,9 +138,10 @@ public final void processRegistrationMessage(final Message<JsonObject> regMsg) {
             final JsonObject body = regMsg.body();
             final String tenantId = body.getString(MessageHelper.APP_PROPERTY_TENANT_ID);
             final String deviceId = body.getString(MessageHelper.APP_PROPERTY_DEVICE_ID);
-            final String action = body.getString(RegistrationConstants.FIELD_ACTION);
+            final String subject = body.getString(MessageHelper.SYS_PROPERTY_SUBJECT);
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/registration/BaseRegistrationService.java]: this is the <em>operation</em>. we only use the `subject` property for conveying it ...

File: core/src/main/java/org/eclipse/hono/util/CredentialsConstants.java [hunk 2]
```diff
@@ -66,74 +57,19 @@ public static JsonObject getCredentialsMsg(final Message message) {
...
-        jsonObject.put(MessageHelper.APP_PROPERTY_TENANT_ID, tenantId);
-        if (deviceId != null) {
-            jsonObject.put(MessageHelper.APP_PROPERTY_DEVICE_ID, deviceId);
-        }
-        jsonObject.put(MessageHelper.APP_PROPERTY_STATUS, Integer.toString(status));
-        if (payload != null) {
-            jsonObject.put(FIELD_PAYLOAD, payload);
-        }
-        return jsonObject;
+        return getServiceRequestAsJson(subject,tenantId,null,payload);
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/CredentialsConstants.java]: spaces after commas

can you fix your code formatter, please?

File: core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    /**
+     * Build a Proton message as a reply for an endpoint from the json message received from the vert.x eventbus
+     * from the implementing service.
+     *
+     * @param endpoint The endpoint the reply message will be built for.
+     * @param message The json message received.
+     * @return Message The built Proton message.
+     */
+    public static final Message getAmqpReply(final String endpoint, final io.vertx.core.eventbus.Message<JsonObject> message) {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java]: it doesn't seem to be very helpful to accept a `io.vertx.core.eventbus.Message` as input param. Why not expect a plain `JsonObject`?

File: core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        final String status = message.body().getString(MessageHelper.APP_PROPERTY_STATUS);
+        final JsonObject correlationIdJson = message.body().getJsonObject(MessageHelper.SYS_PROPERTY_CORRELATION_ID);
+        final Object correlationId = decodeIdFromJson(correlationIdJson);
+        final boolean isApplCorrelationId = message.body().getBoolean(MessageHelper.ANNOTATION_X_OPT_APP_CORRELATION_ID, false);
+        return getAmqpReply(endpoint, status, correlationId, tenantId, deviceId, isApplCorrelationId,
+                message.body().getJsonObject(CredentialsConstants.FIELD_PAYLOAD));
+    }
+
+    /**
+     * Build a Proton message as a reply for an endpoint from the json message received from the vert.x eventbus
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java]: that is not true, this method takes the individual values that have been conveyed in the message received via the event bus. So this should probably be something like:
> Creates an AMQP message containing the reply to a request.

File: core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        if (payload != null) {
+            message.setContentType("application/json; charset=utf-8");
+            message.setBody(new AmqpValue(payload.encode()));
+        }
+        return message;
+    }
+
+    /**
+     * Build a Json object as a reply for internal communication via the vert.x event bus.
+     * Services use this object to build their response when replying to a request that was received for processing.
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java]: `Service implementations may use this method ...`

File: core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+     * @param operation The operation that shall be processed by the service.
+     * @param tenantId The tenant for which the message was processed.
+     * @param deviceId The device that the message relates to.
+     * @return JsonObject The json object for the request that is to be sent via the vert.x event bus.
+     */
+    public static final JsonObject getServiceRequestAsJson(final String operation, final String tenantId, final String deviceId) {
+        return getServiceRequestAsJson(operation, tenantId, deviceId, null);
+    }
+
+//    public static JsonObject getServiceRequestAsJson(final String action, final String tenantId, final JsonObject payload) {
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/RequestResponseApiConstants.java]: remove

File: core/src/main/java/org/eclipse/hono/util/CredentialsConstants.java [hunk 7]
```diff
@@ -8,34 +8,25 @@
...
     public static final String OPERATION_ADD                     = "add";
     public static final String OPERATION_UPDATE                  = "update";
     public static final String OPERATION_REMOVE                  = "remove";
 
     /* message payload fields */
-    public static final String FIELD_PAYLOAD                     = "payload";
-    public static final String FIELD_ENABLED                     = "enabled";
     public static final String FIELD_TYPE                        = "type";
     public static final String FIELD_AUTH_ID                     = "auth-id";
     public static final String FIELD_DEVICE_ID                   = "device-id";
```
Comment (sophokles73) [inline core/src/main/java/org/eclipse/hono/util/CredentialsConstants.java]: this is the same as `RegistrationConstants.FIELD_HONO_ID`. Please pull up and use consistently.

**Comment by sysexcontrol** [comment]
Provided update, please have a look again, thanks!

File: service-base/src/main/java/org/eclipse/hono/service/registration/RegistrationEndpoint.java [hunk 8]
```diff
@@ -90,6 +90,6 @@ protected boolean passesFormalVerification(ResourceIdentifier linkTarget, Messag
 
     @Override
     protected Message getAmqpReply(io.vertx.core.eventbus.Message<JsonObject> message) {
-        return RegistrationConstants.getAmqpReply(RegistrationConstants.REGISTRATION_ENDPOINT,message);
+        return RegistrationConstants.getAmqpReply(RegistrationConstants.REGISTRATION_ENDPOINT,message.body());
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/registration/RegistrationEndpoint.java]: space after comma ;-)

**Comment by sysexcontrol** [comment]
Code formatter/inspection not _quite_ working yet :-1: !@%$
Updated :-)

**Comment by sophokles73** [comment]
ok, can you sqash and rebase?

**Comment by sysexcontrol** [comment]
Done.

**Pull request merged**